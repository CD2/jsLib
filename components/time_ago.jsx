import React from 'react'
import PropTypes from 'prop-types'
import { observable, action, computed } from 'mobx'
import { observer } from 'mobx-react'
import moment from 'moment'

const SECOND = 1000
const MINUTE = SECOND * 60
const HOUR = MINUTE * 60
const DAY = HOUR * 24

@observer
export class TimeAgo extends React.Component {

  static propTypes = {
    time: PropTypes.string.isRequired,
  }

  componentDidMount() {
    this.timeout()
  }

  componentWillUnmount() {
    clearTimeout(this.timer)
  }

  @observable displayTime = ``
  @computed get time() {
    return new Date(this.props.time).getTime()
  }

  getInterval() {
    const diff = Date.now() - this.time

    if (diff < MINUTE) {
      return SECOND*5
    } else if (diff < HOUR) {
      return SECOND*15
    } else if (diff < DAY) {
      return MINUTE*15
    }
    return null
  }

  @action timeout = () => {
    let interval = this.getInterval()
    this.displayTime = moment(new Date(this.props.time)).fromNow()

    if (interval) {
      this.timer = setTimeout(this.timeout, interval)
    }
  }

  render() {
    return (
      <span>{this.displayTime}</span>
    )
  }

}
export default TimeAgo
