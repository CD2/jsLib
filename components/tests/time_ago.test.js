import t from 'utils/test_helper'

import { TimeAgo } from '../time_ago'

t.describe_component(TimeAgo, { time: new Date() }, (c) => {
  describe(`proptypes`, () => {
    c.accepts_prop(`time`)
  })
  describe(`expecting seconds`, () => {
    c.call_function(`getInterval`).expect_return_to_eq(5000)
    c.call_function(`timeout`)
    c.expect_state({ displayTime: `a few seconds ago` })
    c.has_css(`span`).with_text(`a few seconds ago`)
  })

  describe(`expecting minutes`, () => {
    c.with_props({ time: new Date(Date.now() - 60100) }, (c) => {
      c.call_function(`getInterval`).expect_return_to_eq(15000)
      c.call_function(`timeout`)
      c.expect_state({ displayTime: `a minute ago` })
      c.has_css(`span`).with_text(`a minute ago`)
    })
  })
  describe(`expecting hours`, () => {
    c.with_props({ time: new Date(Date.now() - 6001100) }, (c) => {
      c.call_function(`getInterval`).expect_return_to_eq(900000)
      c.call_function(`timeout`)
      c.expect_state({ displayTime: `2 hours ago` })
      c.has_css(`span`).with_text(`2 hours ago`)
    })
  })
})
