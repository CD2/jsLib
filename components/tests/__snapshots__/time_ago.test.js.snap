// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <TimeAgo /> expecting hours Component <TimeAgo /> renders correctly 1`] = `<span />`;

exports[`Component <TimeAgo /> expecting minutes Component <TimeAgo /> renders correctly 1`] = `<span />`;

exports[`Component <TimeAgo /> renders correctly 1`] = `<span />`;
